### PR TITLE
[FEATURE factory-for] Implement factoryFor

### PR DESCRIPTION
**Pull request by chadhietala**
Implements `lookup.factoryFor` to avoid double extending all objects.

See [RFC](https://github.com/emberjs/rfcs/blob/master/text/0150-factory-for.md)

### CODE REVIEW

**Comment by rwjblue** [comment]
Haven't had a chance to review in detail, but I did notice that this still needs feature flagging.

**Comment by chadhietala** [comment]
@stefanpenner I think this good for first round of review. Currently no feature flagging has been done because these are effectively internal changes with no user-land implications.

**Comment by rwjblue** [comment]
> Currently no feature flagging has been done because these are effectively internal changes with no user-land implications.

O_o, isn't the whole point of the RFC to expose this as public API?

**Comment by mixonic** [comment]
@chadhietala 

> inject the deprecated container in this new public API

I'm not quite sure what this means, the container is not deprecated. Will review here.

File: packages/container/lib/container.js
Comment (mixonic) [inline packages/container/lib/container.js]: I expected the implementation here to be

```js
let factory = this._lookupFactory(name);
return factory.create(options);
```

for the initial pass of `factoryFor` `create` per https://github.com/emberjs/rfcs/blob/master/text/0000-factoryFor.md#detailed-design. Additionally https://github.com/emberjs/rfcs/blob/master/text/0000-factoryFor.md#development-mode-proxy. In practice literally implementing as `_lookupFactory(name).create(props)` might not work, but it seems like it should and would preserve the double-extend behavior during instance creation. We should decide if this is really needed, will raise it in some conversations.

File: packages/container/lib/container.js
Comment (mixonic) [inline packages/container/lib/container.js]: putting the faster checks first seems ideal:

```js
function isFactoryClass(container, fullname, { instantiate, singleton }) {
  return (
    (singleton === false || !isSingleton(container, fullname)) &&
    (instantiate === false || !shouldInstantiate(container, fullname))
  );
}
```

thus avoiding the function calls in many cases.

File: packages/container/lib/container.js
Comment (mixonic) [inline packages/container/lib/container.js]: can this function be renamed now, since it is not the `factoryFor` one would normally be looking for here.

File: packages/container/lib/container.js
Comment (mixonic) [inline packages/container/lib/container.js]: So this means`container.lookup` is now based on the the non-double-extended factory? This seems like the behavior we didn't intend to change until `_lookupFactory` is removed.

File: packages/container/lib/container.js
Comment (mixonic) [inline packages/container/lib/container.js]: Must be `until: '2.13'`. 2.12 is an LTS release and must contain this deprecation.

File: packages/container/tests/container_test.js
Comment (mixonic) [inline packages/container/tests/container_test.js]: There is also an `ignoreDeprecation` which might be appropriate/useful here. Example:

https://github.com/emberjs/ember.js/blob/aeeddf7c00b42a340494465252481b5414d87009/packages/ember-runtime/tests/controllers/controller_test.js#L118

File: packages/container/tests/container_test.js
Comment (mixonic) [inline packages/container/tests/container_test.js]: This test shows why we need to keep the "double extend" the default path for `lookup` until at least 2.13.

File: packages/container/lib/container.js
Comment (mixonic) [inline packages/container/lib/container.js]: Would like to have tests included that:

* getting a property that is not `class` or `create` throws an error
* setting a property throws an error
* calling `instanceof` with the return value from `factoryFor` throws an error

Per: https://github.com/emberjs/rfcs/blob/master/text/0000-factoryFor.md#development-mode-proxy

**Review by mixonic — CHANGES_REQUESTED** [review]
Thanks so much for getting this going @chadhietala. The big item is that `lookup` should still be using double extend. If you want to add a second feature flag (ideally in a followup) that experimentally switches `lookup` to the non-double-extended factory, that would definitely be welcome as a distinct feature flag.

File: packages/container/lib/container.js
Comment (chadhietala) [inline packages/container/lib/container.js]: Not sure if I follow. Can we not just transition all the internal stuff over to `factoryFor`?

File: packages/ember-extension-support/lib/data_adapter.js
Comment (mixonic) [inline packages/ember-extension-support/lib/data_adapter.js]: It seems there is some use of the types in Ember-Data. For example [here attributes are read off the class](https://github.com/emberjs/data/blob/c9d8212c857ca78218ad98d11621819b38dba98f/addon/-private/system/debug/debug-adapter.js#L32). Removing the double-extend means the `attributes` CP would be called and memoized on the base class, and thus shared between test runs etc.

This is definitely a change, described [in the design section](https://github.com/emberjs/rfcs/blob/master/text/0000-factoryFor.md#design-of-applicationinstancefactoryfor).

Some other examples of libraries this would impact are listed [in the drawbacks section](https://github.com/emberjs/rfcs/blob/master/text/0000-factoryFor.md#drawbacks).

File: packages/container/lib/container.js
Comment (mixonic) [inline packages/container/lib/container.js]: The intent was to encourage libs/apps to use the new API and drop their current reliance on the double-extended class, then change over internals and any other small breaking changes that will happen after most apps have refactored toward `factoryFor` and stopped expecting a double-extend.

**Comment by chadhietala** [comment]
I made some changes to this, but likely needs further guidance.

**Comment by stefanpenner** [comment]
reviewing now

File: packages/container/lib/container.js
Comment (stefanpenner) [inline packages/container/lib/container.js]: I don't believe we want to mutate the `class`, note: it is perfectly legal to register a class multiple times with different injection rules/similar for each. When this occurs, nothing should collide and currently NAME_KEY would collide. I believe NAME_KEY can be provided merely as a member of `props`. I believe in most cases this should work IFF

https://github.com/emberjs/ember.js/blob/master/packages/ember-runtime/lib/system/core_object.js#L546

changed to:

```js
var ret = '<' + (this[NAME_KEY] || this.constructor.toString()) + ':' + guidFor(this) + extension + '>';
```

This would mean, `class.toString()` wouldn't be terribly useful, but I don't believe it can be accomplished

File: packages/container/lib/container.js
Comment (stefanpenner) [inline packages/container/lib/container.js]: I don't believe we want to mutate the class

File: packages/container/lib/container.js
Comment (stefanpenner) [inline packages/container/lib/container.js]: yes, we should only have 1 behavior in the system at any point in time, once we cut over it should be all no double extend.

File: packages/container/lib/container.js
Comment (stefanpenner) [inline packages/container/lib/container.js]: does this refer to the `factory.class` or our new factory object, if it is our new factory object this may no longer make sense?

File: packages/container/lib/container.js
Comment (stefanpenner) [inline packages/container/lib/container.js]: @mixonic  / @rwjblue should this become a symbol or something in the future?

Comment (mixonic) [inline packages/container/lib/container.js]: In the future perhaps. I don't think it should be changed in this effort.

File: packages/ember-routing/lib/system/generate_controller.js
Comment (stefanpenner) [inline packages/ember-routing/lib/system/generate_controller.js]: if previously, injections where set on `controller:basic` they would now be lost (as we bypass DI if we do `.class`

File: packages/ember-routing/lib/system/route.js
Comment (stefanpenner) [inline packages/ember-routing/lib/system/route.js]: In a future iteration, we should consider deprecating this feature. it is... dubious at best (and was my fault...)

File: packages/ember-routing/lib/system/router.js
Comment (stefanpenner) [inline packages/ember-routing/lib/system/router.js]: similar to the above `controller:basic` example, if injections existed on `route:basic` at this point they would be lost, which I believe is a breaking change.

**Review by stefanpenner — DISMISSED** [review]

File: packages/ember-routing/lib/system/router.js
Comment (chadhietala) [inline packages/ember-routing/lib/system/router.js]: Same question as above.

File: packages/ember-routing/lib/system/generate_controller.js
Comment (mixonic) [inline packages/ember-routing/lib/system/generate_controller.js]: @chadhietala and I went over this. The change as it stands would indeed squash DI in the case where there was an injection on `controller:basic` and you expected it to show up on `controller:generated-controller`. Of course that whole concept, that your DI onto `controller:basic` will show up on `controller:generated-controller`, is pretty WAT. But we should preserve it if possible.

File: packages/ember-routing/lib/system/generate_controller.js
Comment (mixonic) [inline packages/ember-routing/lib/system/generate_controller.js]: `isGenerated` can be killed off.

File: packages/ember-routing/lib/system/router.js
Comment (mixonic) [inline packages/ember-routing/lib/system/router.js]: If we care about injections on `route:basic` being available on all generated routes, we can take the current implementation:

```js
let DefaultRoute = routeOwner._lookupFactory('route:basic');
routeOwner.register(fullRouteName, DefaultRoute.extend());
handler = routeOwner.lookup(fullRouteName);
// Later on may lookup this name again
```

And hack a version that should do the double-DI (DI for `route:basic` and `route:generated-thing` with the current implementation:

```js
let factory = routeOwner.factoryFor('route:basic');
let DefaultRouteProxy = {
  extend(...diArgs) {
    return {
      create(...createArgs) {
         return factory.create(assign({}, ...diArg, ...createArgs));
      }
    };
  },
  create(...createArgs) {
    /* Not sure this is ever going to be used in the current container double-extend implementation */
    return factory.create(assign({}, ...createArgs));
  }
}
```

However in 99.99999% of cases I don't expect someone is doing an injection on `route:basic` (or `controller:basic`), instead they are probably doing injections on the type. To keep these cases fast, we can check the injections list and if possible simple re-register the class:

```js
let DefaultRoute = routeOwner.factoryFor('route:basic').class;
routeOwner.register(fullRouteName, DefaultRoute);
handler = routeOwner.lookup(fullRouteName);
```

In this case DI on the type still works.

File: packages/container/lib/container.js
Comment (mixonic) [inline packages/container/lib/container.js]: It does looks like this should be added to the injections [per master](https://github.com/emberjs/ember.js/blob/e601285bed0e2e108d41a68a61dc2098aac26dd7/packages/container/lib/container.js#L273) instead of set on the class.

File: packages/container/lib/container.js
Comment (mixonic) [inline packages/container/lib/container.js]: This refers to `extend` on the registered class, not `extend` on the factory object. The idea being `extend` is supported, injections have been made.

`instantiate` is only called by `lookup` though, and in that case we want the double extend to continue for now so this test would hold true.

**Comment by mixonic** [comment]
Summary comment of our planning:

* [ ] There should be two feature flags. One adding `factoryFor` with the legacy double-extend implementation, one enabling no-more-double-extend mode.
* [ ] `isGenerated` is no longer needed as a flag on generated classes. (@rwjblue sanity check me?)
* [ ] In the `factoryFor` documentation we should note that by using `factoryFor` the consuming code takes responsibility for destroying the instance.
* [ ] It seems we can solve the injections-on-generated-singletons (routes, controllers) issue by continuing to use the register hack already in master, but we should also implement an above noted meta-hack to keep injections on `route:basic` and `controller:basic` showing up on all generated instances. I actually think this behavior is a) wrong and b) not relied upon and would prefer to just nuke it, but it warrants discussion.
* [ ] `NAME_KEY` needs fixin'
* [ ] `_nameToClass` in the data_adapter seems like it should return either the double extended class or the registered class depending on if the second no-more-double-extend feature flag is enabled.

**Comment by rwjblue** [comment]
> isGenerated is no longer needed as a flag on generated classes. (@rwjblue sanity check me?)

Confirm.

File: packages/ember-routing/lib/system/router.js
Comment (chadhietala) [inline packages/ember-routing/lib/system/router.js]: So this doesn't seem to work for generated controllers due to the fact `.proto()` called right after we generate the controller. https://github.com/emberjs/ember.js/blob/192017b71070d5a32f24ec215ad75dd702d0c0bc/packages/ember-routing/lib/system/route.js#L193

**Comment by chadhietala** [comment]
@mixonic so I think there is some problems with the last 2 items. The `_nameToClass` item above can't actually be tested sense flags are turned on in binary fashion. I can put the guard in there but we are basically flying blind on the `no-double-extend` path. The generated controllers and routes almost work however in the controller case we try to call `proto()` on the generated class so doing the proxying class hack doesn't really work.

File: packages/container/lib/container.js
Comment (chadhietala) [inline packages/container/lib/container.js]: This seems bad but don't know how to make the Node tests pass without doing this.

File: packages/container/lib/container.js
Comment (mixonic) [inline packages/container/lib/container.js]: I think this check should be `ember-factory-for`?

Additionally it seems like you would pass the boolean as an argument to `deprecate`, doesn't really matter though :-p

File: packages/container/lib/container.js
Comment (mixonic) [inline packages/container/lib/container.js]: "of factory" -> "of the factory"

"repsonsible" -> "responsible"

File: packages/container/lib/container.js
Comment (mixonic) [inline packages/container/lib/container.js]: "for the destruction of the factory" -> "for the destruction of any factory instances, as there is no way for the container to ensure instances are destroyed when it itself is destroyed"

File: packages/container/tests/container_test.js
Comment (mixonic) [inline packages/container/tests/container_test.js]: `ignoreDeprecation` may be more appropriate for a test utility method.

File: packages/container/tests/container_test.js
Comment (mixonic) [inline packages/container/tests/container_test.js]: yeah `ignoreDeprecation` would allow you to not have this change in expect count for an unrelated test.

File: packages/ember-application/tests/system/dependency_injection/default_resolver_test.js
Comment (mixonic) [inline packages/ember-application/tests/system/dependency_injection/default_resolver_test.js]: It is concerning that this test coverage is lost when the feature flag is enabled?

File: packages/ember-glimmer/lib/syntax/render.js
Comment (mixonic) [inline packages/ember-glimmer/lib/syntax/render.js]: Doesn't using factoryFor here mean the double extended class is never used for `render`, regardless of the feature flag setting?

**Review by mixonic — CHANGES_REQUESTED** [review]

**Review by mixonic — APPROVED** [review]

**Comment by mixonic** [comment]
OMG this is good to go. To merge:

* One of us will rebase this down to a more meaningful commit history with `[FEATURE factory-for]` in the commits
* I would like to boot the build with various flags enabled/disabled in a real world app before merge. Just a quick sanity check.

Excited :-D

**Comment by chadhietala** [comment]
Quick benchmarks from Travis-Web. This is consistent across runs and just looking at the homepage that is likely pretty static.

__Before: (Prod / Feature Flags Off)__
screen shot 2016-12-17 at 7 31 51 pm

__After: (Prod / Feature Flags On)__
screen shot 2016-12-17 at 7 32 51 pm

**Comment by rwjblue** [comment]
@chadhietala - I assume the `After: (Prod / Feature Flags On)` screenshots above are with both `ember-factory-for` and `ember-no-double-extend` flags on, is that right?  Can you test with only `ember-factory-for` (I'm mostly looking to try to figure out if we will have temporary regressions in 2.12)?

File: packages/ember-application/lib/system/engine-instance.js
Comment (rwjblue) [inline packages/ember-application/lib/system/engine-instance.js]: Lets make this part of the `ContainerProxyMixin`, this would simplify issues in `ember-test-helpers` land and actually keep the logic in the correct place (currently `.lookup` and `._lookupFactory` are done in `ContainerProxyMixin`).

File: packages/ember-application/lib/system/engine-instance.js
Comment (rwjblue) [inline packages/ember-application/lib/system/engine-instance.js]: This can be moved to the mixin also...

**Review by rwjblue — DISMISSED** [review]

File: packages/container/lib/container.js
Comment (rwjblue) [inline packages/container/lib/container.js]: Can we move this into `ember-utils` and export as `HAS_NATIVE_PROXY`?  Very similar to what was done in https://github.com/emberjs/ember.js/pull/14649 for `packages/ember-utils/lib/weak-map-utils.js`.

File: tests/node/helpers/build-owner.js
Comment (rwjblue) [inline tests/node/helpers/build-owner.js]: See comments in https://github.com/emberjs/ember.js/pull/14360#pullrequestreview-13465262 and https://github.com/emberjs/ember-test-helpers/pull/191, we should be able to remove this.

**Comment by mixonic** [comment]
I've done some benchmarks against todomvc. I would expect the impact on a larger app to be much more significant (as @chadhietala's profiling shows).

This PR may have a minor performance regression:

screen shot 2016-12-18 at 2 19 48 pm

The `ember-factory-for` feature flag does not seem to have a significant impact on performance (which is expected).

However the `ember-no-double-extend` feature flag makes a significant improvement.

screen shot 2016-12-18 at 2 22 14 pm

I think we can continue to explore issues around backwards compatibility, performance, and the migration path to `factoryFor` on master. Merging this.

@chadhietala Thanks much for your patience and push on this gnarly, complex, cross-cutting part of Ember's API. The payoff (especially in big apps) is going to be really notable.

**Comment by stefanpenner** [comment]
👍

**Comment by sandstrom** [comment]
Awesome! Nice to be able to move into a public API for this + great performance improvements! ⛵️

**Pull request merged**
